### PR TITLE
Enhance Admin Account Manager Features

### LINKED ISSUE #110: 【SB-03】Manage Subscription
### Overview

Infra:
Management of the free plan (5GB) and paid plan (200GB) (tentative).

Admin:
User Subscriptions Managing

- #116

User:
Plan changes, upgrades, and downgrades.
Payment Method Changes

### References

There are some tasks related to this requirements.
- #72 
- #73 
- #74 

### Checklist

- [x] Add storage and subscription information to account management page

Admin users should be able to 
- [x] See all users data (does this refresh correctly to reflect actual s3 folder size?)
- [x] delete user data (Maybe can be done after https://github.com/arayabrain/optinist-for-cloud/issues/122)
- [x] delete users
- [x] add users
- [x] edit subscription details
- [x] edit subscription expiry date
- [x] edit storage quota

### PR DESCRIPTION
**Pull request by tsuchiyama-araya**
## Summary

  Admin capability to edit user subscription details (plan, expiration date, storage quota) directly from the Account Manager page. This enables admins to manage subscriptions for edge cases like trial extensions, manual
  plan corrections, and storage quota adjustments.

  ## Changes

  ### Backend

  - **New endpoint**: `PUT /admin/users/{user_id}/subscription` — Admin-only endpoint to update a user's subscription plan, expiration date, and storage quota
  - **New schema**: `UserSubscriptionUpdate` in `schemas/users.py` with Pydantic validation (`storage_quota_bytes > 0`, `reason` non-empty)
  - **New CRUD function**: `update_user_subscription_admin()` in `crud_users.py` — Directly updates `UserSubscription` and `UserStorageUsage` tables (bypasses Stripe intentionally)
  - **Audit logging**: Every change is recorded in `subscription_audit_log` with old/new values, admin ID, and reason
  - **New DB migration**: `l901l9310026_add_subscription_audit_log.py` creates the audit log table
  - **New response field**: Added `subscription_expiration` to the `User` schema for frontend display

  ### Frontend

  - **New modal**: `SubscriptionEditModal` component in Account Manager with:
    - Plan selector dropdown (Free / Premium)
    - Expiration date picker (disabled for Free plan)
    - Storage quota input (in GB, minimum 1)
    - Current storage usage display
    - Warning alert that changes bypass Stripe
    - Reason field (required) for audit trail
  - **New action column button**: Credit card icon on each user row to open the subscription edit modal
  - **Redux integration**: New `updateUserSubscription` async thunk with loading state management

  ### Tests

  - **New test file**: `test_users_admin_subscription.py` — 14 tests covering:
    - Schema validation: valid inputs, required fields, type checking, boundary values (0, negative, empty string)
    - Audit snapshot serialization

  ## Validation Rules

  ### Schema validation (Pydantic — automatic 422 response)

  | Field | Type | Required | Constraints |
  |---|---|---|---|
  | `plan_id` | `int` | Yes | Must be parseable as integer |
  | `expiration` | `datetime` (ISO 8601) | No | Valid ISO 8601 string if provided; `null`/omitted accepted |
  | `storage_quota_bytes` | `int` | Yes | Must be `> 0` |
  | `reason` | `str` | Yes | Must be non-empty (min length 1) |

  ### Business logic validation (returns 400/404)

  | Condition | Status | Detail |
  |---|---|---|
  | User not found or not in admin's organization | 404 | `"User not found"` |
  | `plan_id` not in `{1 (Free), 2 (Premium)}` | 400 | `"Invalid plan_id: {value}"` |
  | User has no `subscription_users` record | 400 | `"User has no subscription record"` |
  | User has no `user_storage_usage` record | 400 | `"User has no storage record"` |

  Note: If `expiration` is `null`, it defaults to `datetime.now(UTC)`. This is intentional for Free plan downgrades where expiration is not meaningful.

  ### Frontend UI validation

  | Field | Validation | Mechanism |
  |---|---|---|
  | Plan | Constrained to Free / Premium dropdown | `<Select>` with fixed `<MenuItem>` values |
  | Expiration | Required for Premium (Save disabled if empty); omitted for Free | Button `disabled` prop |
  | Storage Quota | Minimum 1 GB, integer only | `Math.max(1, Math.floor(...))` in onChange + `inputProps={{ min: 1 }}` |
  | Reason | Required, whitespace-only rejected | `!reason.trim()` check on Save button |

  ### Known accepted gaps

  - **No upper bound** on `storage_quota_bytes` — admins may legitimately need large quotas
  - **No future-date check** on `expiration` — admins may need to set past dates to immediately expire a subscription
  - **No max length** on `reason` — stored as `TEXT` in DB, no practical limit needed

  ## Additional behavior

  - **`scheduled_downgrade` reset**: When an admin manually changes a subscription, any pending Stripe-initiated downgrade flag is cleared to prevent conflicts
  - **`db.rollback()`**: Both error paths (`HTTPException` and unexpected exceptions) roll back the transaction
  - **Expiration time**: Frontend sends end-of-day (`T23:59:59Z`) so the subscription remains valid through the selected date
  - **Error logging**: All validation rejections are logged at WARNING level with user ID and detail; unexpected errors at ERROR level with stack trace

  ## Test plan

  - [x] As admin, open Account Manager and verify the credit card icon button appears on each user row
  - [x] Click the button and verify the modal opens with the user's current subscription details pre-populated
  - [x] Change plan from Free to Premium, set expiration date, adjust storage quota, and save
  - [x] Verify the user list refreshes with updated subscription status and storage quota
  - [x] Verify the warning about bypassing Stripe is displayed in the modal
  - [x] Verify non-admin users cannot access the endpoint (403)
  - [x] Verify invalid plan_id values are rejected (400)
  - [x] Verify `storage_quota_bytes: 0` and negative values are rejected (422)
  - [x] Verify empty `reason` is rejected (422)
  - [x] Verify audit log entry is created with correct old/new values

### CODE REVIEW

File: frontend/src/pages/AccountManager/index.tsx [hunk 1]
```diff
@@ -12,20 +12,27 @@ import { useNavigate, useSearchParams } from "react-router-dom"
 
 import { useSnackbar, VariantType } from "notistack"
```
Comment (itutu-tienday) [inline frontend/src/pages/AccountManager/index.tsx:13]: Have you considered the following?
(Considerations for preventing and detecting unauthorized or abnormal update operations)
If you haven't, it might be a good idea to take a look.

https://github.com/arayabrain/araya-optinist/issues/110#issuecomment-4045347961

> > A few things can go wrong.
>
> - It would be good to incorporate specifications to detect and cover those anticipated problems (important and relatively easy to implement).
> - To enable investigation and auditing, it's a good idea to record information about users who have manually updated their subscriptions.
> - (such as by recording the operation in the log table or by displaying a mark in the user list indicating that the subscription has been manually updated).

First, consider a method that can be implemented without incurring much effort.

Comment (tsuchiyama-araya) [reply frontend/src/pages/AccountManager/index.tsx:13]: For this I will create a sheet for listing the errors happened and what changes had been done with the user.
I think this is less work than to implement it and testing.

Comment (itutu-tienday) [reply frontend/src/pages/AccountManager/index.tsx:13]: Does this mean that work execution history will be manually entered and managed by people in a spreadsheet?
If so, there are concerns about the unreliability of the records and the possibility of the operation becoming a mere formality.

It is thought that some kind of function, even a simple one, should be used to address this.
→ There is a risk that evidence will not be able to be verified when problems arise regarding the user agreement.
→ It would be good to request a review from AI, referring to the method proposed in #110.

*If my points contradict the premise (e.g., wanting to expedite the release), please provide further details.

Comment (tsuchiyama-araya) [reply frontend/src/pages/AccountManager/index.tsx:13]: @itutu-tienday 

This concern is  addressed in the current implementation. Subscription changes made through the
Account Manager are not manually tracked — they are automatically recorded in a subscription_audit_log    
table at the database level.                                                                              
                                                                                                            
Why not a spreadsheet?                                                                                    
  - Human error — entries can be forgotten, mistyped, or inconsistent
  - No enforcement — nothing prevents someone from skipping the entry
  - No link to the actual change — there's no guarantee the spreadsheet matches what really happened in the
  database
  - Difficult to audit — searching, filtering, and verifying records across a spreadsheet is unreliable at
  scale
  - Tampering risk — spreadsheet entries can be edited or deleted after the fact without trace

How our implementation works instead:
  - When an admin edits a user's subscription, the system automatically captures the old values and new
  values (plan, expiration, storage quota) as JSON
  - The admin is required to enter a reason — the Save button is disabled until a reason is provided
  - The audit record is written in the same DB transaction as the subscription change, so it's impossible to
   make a change without producing a log entry
  - Each record includes: who was changed, who made the change, what changed (before/after), why, and when

Regarding the specific concerns:
  - Unreliable records: The old/new values are captured programmatically by the system, not entered manually
   — there is no room for human error or omission
  - Operation becoming a formality: The reason field is enforced at both the frontend (button disabled) and
  backend (field required in schema) level
  - Evidence verification: The exact diff of every change is stored with timestamps and admin identity,
  providing verifiable evidence if issues arise

As a future enhancement, we could add an audit log viewer in the admin UI or integrate AI review of reason
   quality (as suggested in #110), but the automated audit trail is already functional in this release.

File: studio/app/common/core/users/crud_users.py [hunk 2]
```diff
@@ -571,8 +572,36 @@ async def update_user_subscription_admin(
         )
         if storage is None:
             raise HTTPException(status_code=400, detail="User has no storage record")
+
+        # Capture old values before applying changes
+        old_value = {
```
Comment (itutu-tienday) [inline studio/app/common/core/users/crud_users.py]: Defining the data to be recorded using a dataclass or similar method, rather than a raw dictionary, would likely improve the maintainability of the code.

File: studio/app/common/core/users/crud_users.py [hunk 3]
```diff
@@ -530,10 +531,12 @@ async def update_user_subscription_admin(
     user_id: int,
     data: UserSubscriptionUpdate,
     organization_id: int,
+    admin_user_id: int,
```
Comment (itutu-tienday) [inline studio/app/common/core/users/crud_users.py]: Passing two separate user information parameters (organization_id, user_id) is redundant.

Passing just one parameter (User model) indicating the "updater" should be sufficient.

File: frontend/src/pages/AccountManager/index.tsx [hunk 4]
```diff
@@ -284,6 +296,136 @@ const ModalComponent = ({
...
+              label="Plan"
+              onChange={(e) => setPlanId(Number(e.target.value))}
+            >
+              <MenuItem value={PLAN_FREE}>Free</MenuItem>
+              <MenuItem value={PLAN_PREMIUM}>Premium</MenuItem>
+            </MuiSelect>
+          </FormControl>
+
+          <TextField
+            label="Expiration Date"
```
Comment (itutu-tienday) [inline frontend/src/pages/AccountManager/index.tsx]: When I selected the free plan, the date is a required field, but is it really necessary to enter it?

Comment (itutu-tienday) [inline frontend/src/pages/AccountManager/index.tsx]: Is it as expected that we can enter the time as well?

Comment (itutu-tienday) [inline frontend/src/pages/AccountManager/index.tsx]: As mentioned above, the validation rule specifications (including those for fields other than date fields) and the test results are unclear.

Please provide information to clarify the above.  (Enhance the PR explanation)

File: frontend/src/pages/AccountManager/index.tsx [hunk 5]
```diff
@@ -284,6 +296,136 @@ const ModalComponent = ({
...
+        <ButtonModal>
+          <Button variant="outlined" onClick={onClose}>
+            Cancel
+          </Button>
+          <Button
+            variant="contained"
+            disabled={isSubmitting || !expiration || !reason.trim()}
+            onClick={handleSubmit}
+          >
+            Save
```
Comment (itutu-tienday) [inline frontend/src/pages/AccountManager/index.tsx:440]: Local testing yielded the following results:

- The banner "Subscription update failed!" is displayed (appears to be a 400 error).
- No error information appears to be output to the backend (only the following is displayed):
  > INFO: [uvicorn.access] (pid:3318) (task:local) (client:4e782f44eac52685) send():478 - 127.0.0.1:60617 - "PUT /admin/users/1/subscription HTTP/1.1" 400

While I could debug this, the lack of error information being logged to the backend is also a problem. Could you please check the validity of your error handling?

File: studio/app/common/core/users/crud_users.py (hunk 2 above)
Comment (tsuchiyama-araya) [reply studio/app/common/core/users/crud_users.py]: This is already handled. I have a SubscriptionAuditSnapshot Pydantic model in schemas/users.py that defines the typed structure for both old_value and new_value (with plan_id, expiration, and storage_quota_bytes fields). In crud_users.py, snapshots are constructed through this model and then converted to dict via .dict() only at the point of storage, since the DB column is JSON type and needs a plain dict for SQLAlchemy serialization.

File: studio/app/common/core/users/crud_users.py (hunk 3 above)
Comment (tsuchiyama-araya) [reply studio/app/common/core/users/crud_users.py]: Agreed I fixed it. The function now takes admin_user. 
User as a single parameter and derives admin_user.organization.id and admin_user.id from it internally.

File: frontend/src/pages/AccountManager/index.tsx (hunk 4 above)
Comment (tsuchiyama-araya) [reply frontend/src/pages/AccountManager/index.tsx]: Im sorry, I also thought of that after reading this comment.
Therefore, I changed the code when its free plan there is no need to input date.
It will automatically, register the date of input.

Comment (tsuchiyama-araya) [reply frontend/src/pages/AccountManager/index.tsx]: No, I fixed this.
Thank you

Comment (tsuchiyama-araya) [reply frontend/src/pages/AccountManager/index.tsx]: I updated the PR Description. Can you check if the description can fix your questions.

File: frontend/src/pages/AccountManager/index.tsx (hunk 5 above)
Comment (tsuchiyama-araya) [reply frontend/src/pages/AccountManager/index.tsx:440]: I updated the PR Description. Can you check if the description can fix your questions.

File: frontend/src/pages/AccountManager/index.tsx (hunk 1 above)
Comment (itutu-tienday) [reply frontend/src/pages/AccountManager/index.tsx:13]: This update (recording the update history in the database) appears to be an appropriate response.

File: frontend/src/pages/AccountManager/index.tsx (hunk 5 above)
Comment (itutu-tienday) [reply frontend/src/pages/AccountManager/index.tsx:440]: I've confirmed that the cause of the error is now being logged.

The error was "User has no subscription_users record," which seems to apply to almost all newly created accounts (Free users).

Therefore, it seems highly likely that updating a Free user account will result in an error.
Does this contradict the requirements?
(Is this feature primarily for updating Premium users?)

File: frontend/src/pages/AccountManager/index.tsx [hunk 6]
```diff
@@ -284,6 +296,150 @@ const ModalComponent = ({
...
+            type="date"
+            value={isFree ? "" : expiration}
+            onChange={(e) => setExpiration(e.target.value)}
+            InputLabelProps={{ shrink: true }}
+            fullWidth
+            disabled={isFree}
+          />
+
+          <TextField
+            label="Storage Quota (GB)"
```
Comment (itutu-tienday) [inline frontend/src/pages/AccountManager/index.tsx:395]: The PR description states that "Storage Quota has no upper limit," but to prevent editor errors, some kind of validation for the upper limit would be beneficial.

Even simply specifying `maxlength` (5 digits, <10TB) would be safer.

**Comment by itutu-tienday** [comment]
With the implementation up to this point, we're pretty much finished, but I'll add an AI review comment as a final step.

There may be some things that have already been recognized but whose actions have been intentionally omitted; please comment if that's the case.

----

## Code Review

### Findings

#### 1. [Critical] `onSubmit` return value is ignored — modal closes even on failure

**File:** `frontend/src/pages/AccountManager/index.tsx` L336-350

In commit `a5ac66d4`, the `onSubmit` return type was changed to `Promise<boolean>`, but the modal's `handleSubmit` never checks the return value. When the API call fails and `onSubmit` returns `false`, `onClose()` is still called and the modal closes.

```tsx
// Current code
const handleSubmit = async () => {
    setIsSubmitting(true)
    try {
      await onSubmit({...})
      onClose()    // ← reached even when onSubmit returns false
    } finally {
      setIsSubmitting(false)
    }
  }
```

**Suggested fix:**

```tsx
const handleSubmit = async () => {
    setIsSubmitting(true)
    try {
      const success = await onSubmit({...})
      if (success) {
        onClose()
      }
    } finally {
      setIsSubmitting(false)
    }
  }
```

---

#### 2. [Medium] Broken Alembic migration chain

**File:** `studio/alembic/versions/m012m0421037_add_subscription_audit_log.py`

The new migration `m012m0421037` declares `down_revision = "l901l9310026"`, but the old migration file `l901l9310026` was deleted in commit `df4238ada`.

- **Existing environments:** The old migration has already been applied, and the new migration's `if "subscription_audit_log" not in existing_tables` guard makes it idempotent — so it works.
- **Fresh environments:** The file for `l901l9310026` does not exist, so Alembic cannot resolve the chain and will fail.

**Suggested fix (choose one):**

- Restore the `l901l9310026` file with `pass` in both upgrade/downgrade (keep it as a no-op placeholder)
- Change the new migration's `down_revision` directly to `l901l9310026`'s predecessor (`33e781982125`)

---

#### 3. [Medium] No foreign key constraints on `subscription_audit_log`

**Files:** `studio/alembic/versions/m012m0421037_add_subscription_audit_log.py`, `studio/app/common/models/subscription.py`

`user_id` and `changed_by` are defined as `BIGINT` without foreign key constraints referencing the `users` table.

For audit logs, omitting FK constraints is a reasonable design choice — audit records should survive user deletion. However, **this intent should be documented** in the model or migration to prevent future developers from treating it as an oversight.

**Suggested fix:** Add a clarifying comment to the model

```python
user_id: int = Field(
    sa_column=Column(BIGINT, nullable=False, index=True),
    description="The user whose subscription was changed. "
    "No FK constraint — audit records must survive user deletion.",
)
```

---

#### 4. [Low] Modal state may not reset when reopened for a different user

**File:** `frontend/src/pages/AccountManager/index.tsx` L319-331

`useState` initial values are only evaluated on first mount. The current implementation sets `subscriptionEditUser` to `null` on close, which effectively forces a remount. However, React's reconciliation could potentially reuse the component instance, leaving stale state from a previously selected user.

**Suggested fix:** Add a `key` prop to guarantee remounting

```tsx
<SubscriptionEditModal
  key={subscriptionEditUser?.id}   // forces remount per user
  open={subscriptionEditOpen}
  user={subscriptionEditUser}
  onClose={() => {
    setSubscriptionEditOpen(false)
    setSubscriptionEditUser(null)
  }}
  onSubmit={handleSubscriptionEdit}
/>
```

---

#### 5. [Low] Implicit `undefined` return path in `handleSubscriptionEdit`

**File:** `frontend/src/pages/AccountManager/index.tsx` (handleSubscriptionEdit)

If the `dispatch` result is neither `rejected` nor `fulfilled` (e.g., aborted), the function implicitly returns `undefined`. While this is unlikely in practice, adding a fallback `return false` at the end ensures type safety.

```tsx
const handleSubscriptionEdit = async (
    data: UpdateUserSubscriptionDTO,
  ): Promise<boolean> => {
    if (!subscriptionEditUser?.id) return false
    const result = await dispatch(
      updateUserSubscription({...}),
    )
    if (isRejectedWithValue(result)) {
      handleClickVariant("error", "Subscription update failed!")
      return false
    } else {
      handleClickVariant("success", "Subscription updated successfully!")
      return true
    }
    // Unreachable in current logic, but consider adding: return false
  }
```

---

#### 6. [Low] `T23:59:59Z` timezone consideration for expiration

**File:** `frontend/src/pages/AccountManager/index.tsx` L343

```tsx
new Date(expiration + "T23:59:59Z").toISOString()
```

The `Z` suffix sets end-of-day in UTC. If an admin selects a date while in JST, UTC 23:59:59 corresponds to JST 08:59:59 the following day. If the specification explicitly defines this as "UTC 23:59:59", this is correct. If the intent is "end-of-day in the admin's local timezone", this needs adjustment.

---

#### 7. [Low] Missing automated tests for CRUD and endpoint layers

**File:** `studio/tests/app/common/routers/test_users_admin_subscription.py`

The current tests only cover schema validation (`UserSubscriptionUpdate`, `SubscriptionAuditSnapshot`). The following automated tests are missing:

- Unit tests for `update_user_subscription_admin()` with a mock DB (happy path and error cases)
- Integration tests for the endpoint (403 for non-admin, 404 for missing user, 400 for invalid plan_id)
- Verification that audit log entries are created with correct old/new values

Manual testing has been completed per the test plan, but automated tests are recommended to prevent regressions.

---

### Summary

| Severity | # | Finding | Action |
|---|---|---|---|
| Critical | 1 | Modal closes on API failure | Check `onSubmit` return value before calling `onClose` |
| Medium | 2 | Broken Alembic migration chain | Fix `down_revision` or restore deleted migration file |
| Medium | 3 | No FK constraints (undocumented intent) | Add clarifying comments to model |
| Low | 4 | Modal state not guaranteed to reset | Add `key` prop |
| Low | 5 | Implicit `undefined` return | Add fallback `return false` |
| Low | 6 | UTC timezone for expiration | Confirm specification intent |
| Low | 7 | Insufficient automated tests | Add CRUD and endpoint-level tests |

File: frontend/src/pages/AccountManager/index.tsx (hunk 5 above)
Comment (tsuchiyama-araya) [reply frontend/src/pages/AccountManager/index.tsx:440]: Im sorry, but this only happened on my local since there is a old account that was created before implementing the user creation where when the user is created it will register on users and subscription_users.

Comment (tsuchiyama-araya) [reply frontend/src/pages/AccountManager/index.tsx:440]: This edit also works with free user.

File: frontend/src/pages/AccountManager/index.tsx (hunk 6 above)
Comment (tsuchiyama-araya) [reply frontend/src/pages/AccountManager/index.tsx:395]: I updated this to be have a maximum of 10TB

**Comment by tsuchiyama-araya** [comment]
@itutu-tienday 

This is my response for the AI Review.
                                                                                                                  
  1. [Critical] onSubmit return value ignored — Fixed                                                                                                                                                                       
  Good catch. Applied the suggested fix — handleSubmit now checks the boolean return and only calls onClose() on success.                                                                                                     
                                                                                                                                                                                                                              
  2. [Medium] Broken Alembic migration chain — Invalid                                                                                                                                                                        
  The file l901l9310026_add_expiration_deletion.py still exists in studio/alembic/versions/. It was not deleted in commit df4238ada — that commit only updated the migration content. The chain is intact.                    
                                                                                                                                                                                                                              
  3. [Medium] No FK constraints on subscription_audit_log — Fixed                                                                                                                                                             
  Agreed that the intent should be documented. Added clarifying descriptions to both user_id and changed_by fields noting that FK constraints are intentionally omitted so audit records survive user deletion.               
                                                                                                                                                                                                                              
  4. [Low] Modal state may not reset — No action needed                                                                                                                                                                       
  The SubscriptionEditModal is conditionally rendered with {subscriptionEditOpen && subscriptionEditUser && (<SubscriptionEditModal .../>)}. Setting subscriptionEditUser to null on close unmounts the component entirely.   
  When a new user is selected, it's a fresh mount with fresh useState initializers. A key prop is unnecessary here.                                                                                                           
                  
  5. [Low] Implicit undefined return path — No action needed                                                                                                                                                                  
  The if/else is exhaustive — isRejectedWithValue(result) returns a boolean, so one branch always executes. There is no code path that reaches the end of the function without returning. Adding a dead return false after the
   else block would be misleading.                                                                                                                                                                                            
                  
  6. [Low] T23:59:59Z timezone — No action needed                                                                                                                                                                             
  UTC is intentional. The backend stores and evaluates expiration in UTC. Using the admin's local timezone would introduce inconsistency depending on who sets the value.
                                                                                                                                                                                                                              
  7. [Low] Missing automated tests — Acknowledged                                                                                                                                                                             
  Agreed. But I think this should be on different task.

File: frontend/src/pages/AccountManager/index.tsx (hunk 5 above)
Comment (itutu-tienday) [reply frontend/src/pages/AccountManager/index.tsx:440]: So, for users newly created with the latest version of the application, the subscription_users record and other necessary information are all prepared, right?

If so, I think there shouldn't be any problems.

**Review by itutu-tienday — APPROVED** [review]

File: frontend/src/pages/AccountManager/index.tsx (hunk 5 above)
Comment (tsuchiyama-araya) [reply frontend/src/pages/AccountManager/index.tsx:440]: Yes it is.
